torchmultimodal/modules/layers/quantisation.py

```diff
@@ -0,0 +1,73 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+from torch import nn, Tensor
+
+
+class Quantisation(nn.Module):
+    """Quantisation provides an embedding layer that takes in a collection of flattened vectors, usually the
+    output of an encoder architecture, and performs a nearest-neighbor lookup in the embedding space.
+
+    Vector quantisation was introduced in Oord et al. 2017 (https://arxiv.org/pdf/1711.00937.pdf) to generate high-fidelity
+    images, videos, and audio data.
+
+    Args:
+        num_embeddings (int): the number of vectors in the embedding space
+        embedding_dim (int): the dimensionality of the embedding vectors
+    """
+
+    def __init__(self, num_embeddings: int, embedding_dim: int):
+        super().__init__()
+        self.embedding_dim = embedding_dim
+        self.num_embeddings = num_embeddings
+
+        self.embedding = nn.Embedding(self.num_embeddings, self.embedding_dim)
+        self.embedding.weight.data.uniform_(
+            -1 / self.num_embeddings, 1 / self.num_embeddings
+        )
+        self.quantised_vectors = None
+
+    def forward(self, x: Tensor):
+        # Rearrange from batch x channel x n dims to batch x n dims x channel
+        new_dims = (0,) + tuple(range(2, len(x.shape))) + (1,)
+        x_permuted = x.permute(new_dims).contiguous()
+        permuted_shape = x_permuted.shape
+
+        # Flatten input
+        x_flat = x_permuted.view(-1, permuted_shape[-1])
+        # channel dimension should be embedding dim so that each element in encoder
+        # output volume gets associated with single embedding vector
+        assert (
+            x_flat.shape[-1] == self.embedding_dim
+        ), f"Expected {x_flat.shape[-1]} to be embedding size of {self.embedding_dim}"
+
+        # Calculate distances from each encoder output vector to each embedding vector, ||x - emb||^2
+        distances = torch.cdist(x_flat, self.embedding.weight, p=2.0) ** 2
+
+        # Encoding - select closest embedding vectors
+        encoding_indices = torch.argmin(distances, dim=1)
+
+        # Quantise
+        quantised_permuted = self.embedding(encoding_indices).view(permuted_shape)
+
+        # Straight through estimator
+        quantised_permuted = x_permuted + (quantised_permuted - x_permuted).detach()
+
+        # Rearrange back to batch x channel x n dims
+        old_dims = (0,) + (len(x.shape) - 1,) + tuple(range(1, len(x.shape) - 1))
+        quantised = quantised_permuted.permute(old_dims).contiguous()
+        self.quantised_vectors = quantised
```

Comment:
>the quantised vectors are used for commitment, and that's the chosen embedding vectors based on the encoder output. it is different from the embedding itself, which can only be updated via VQ loss / codebook loss

In the forward pass, it doesn't matter. But the commitment loss shouldn't compute the gradient against `quantised_vectors` and that's why the latter needs to be "plucked" out of the graph (or `.detached`). 

Here you are caching it after it has been updated.

https://github.com/openai/jukebox/blob/08efbbc1d4ed1a3cef96e08a931944c8b4d63bb3/jukebox/vqvae/bottleneck.py#L170

Reply:
yes I believe this aligns with the openai code both on the straight through estimator and the commitment loss

Reply:
Noob question: where do we intend to get the second term in the paper's equation (3) from (the VQ objective)? Seems as written it could come from `CommitmentLoss` with the args reversed and `commitment_cost=1.0`, but based on the naming in that module I'm inferring that we don't intend to use it that way.

Reply:
could either implement as two separate loss modules, commitment and codebook, or one single module, vqvae loss. can't imagine a scenario you would want to use them separately, and beta / commitment_cost seems to only apply to commitment loss

Reply:
To prevent the commit loss from taking a gradient against the quantised vector, the latter needs to be cached **before** it's updated by the straight through step:

```
quantised_flat = self.embedding(encoding_indices)
self.quantised_vector = self._postprocess(quantise_flat.detach(), permuted_shape)
quantised_flat = x_flat + (quantised_flat - x_flat).detach()

return quantised_flat
```